subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java

```diff
@@ -187,25 +221,54 @@ private String getFilename(Path dir) {
 
         @Override
         public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
-            String internedName = intern(file.getFileName().toString());
-            if (shouldVisit(file, internedName, false, attrs, builder.getRelativePath())) {
-                builder.visitFile(snapshotFile(file, internedName, attrs));
+            if (attrs.isSymbolicLink()) {
+                BasicFileAttributes targetAttributes = readAttributesOfSymlinkTarget(file, attrs);
+                if (targetAttributes.isDirectory()) {
+                    try {
+                        // TODO: Detect cycles
+                        Path targetDir = file.toRealPath();
+                        symbolicLinkMappings.addFirst(new SymbolicLinkMapping(file.toString(), targetDir.toString()));
+                        Files.walkFileTree(targetDir, EnumSet.noneOf(FileVisitOption.class), Integer.MAX_VALUE, this);
```

Comment:
This recursion looks dangerous. Can we do this with a queue instead perhaps?

Reply:
I added recursion detection. We can't really use a queue, since the visiting needs to happen for the same level.